crates/compiler/builtins/bitcode/src/list.zig

```diff
@@ -140,11 +126,111 @@ pub const RocList = extern struct {
         return new_list;
     }
 
+    // We follow roughly the [fbvector](https://github.com/facebook/folly/blob/main/folly/docs/FBVector.md) when it comes to growing a RocList.
+    // Here is [their growth strategy](https://github.com/facebook/folly/blob/3e0525988fd444201b19b76b390a5927c15cb697/folly/FBVector.h#L1128) for push_back:
+    //
+    // (1) initial size
+    //     Instead of growing to size 1 from empty, fbvector allocates at least
+    //     64 bytes. You may still use reserve to reserve a lesser amount of
+    //     memory.
+    // (2) 1.5x
+    //     For medium-sized vectors, the growth strategy is 1.5x. See the docs
+    //     for details.
+    //     This does not apply to very small or very large fbvectors. This is a
+    //     heuristic.
+    //
+    // In our case, we exposed allocate and reallocate, which will use a smart growth stategy.
+    // We also expose allocateExact and reallocateExact for case where a specific number of elements is requested.
+
+    // calculateCapacity should only be called in cases the list will be growing.
+    // requested_length should always be greater than old_capacity.
+    inline fn calculateCapacity(
+        old_capacity: usize,
+        requested_length: usize,
+        element_width: usize,
+    ) usize {
+        var new_capacity: usize = 0;
+        // TODO: I think this shouldn't be done here.
+        // We should have a smarter way to deal with this that doesn't lead
+        // to allocating a refcount for zero width elements.
```

Comment:
Hmm but would this add some amount of logic to the "normal" cases too?

In practice, zero-size values only come up when dealing with effects. And effects are usually doing something very slow like calling out to a network, filesystem, etc. So an extra allocation is insignificant by comparison. Might not be worth paying a cost elsewhere for that.

Of course if there is no cost elsewhere then great. That's the challenge I guess.

Reply:
That's a really good point. I guess this is better to handle here rather than elsewhere. This probably shouldn't be worth checking in the normal. That being said, I am still very confused why the llvm backend seemed to work without this change. That would suggest to me that the llvm backend is already doing something special for zero width element types.